File: frontend/farm-management-web/src/app/core/labor/attendance.models.ts
import { PagedResponse } from "../models/paged-response.model";

export type AttendanceType =
  | "FULL_DAY"
  | "HALF_DAY"
  | "HOURLY"
  | "NOT_WORKED";

export type AttendanceStatus = "DRAFT" | "FINALIZED";

export interface AttendanceEligibleWorker {
  readonly workerId: string;
  readonly displayName: string;
  readonly firstName: string;
  readonly lastName?: string | null;
  readonly gender: string;
  readonly employmentType: string;
  readonly mobileNumber?: string | null;
  readonly laborCategoryId?: string | null;
  readonly laborCategoryName?: string | null;
  readonly contractorId?: string | null;
  readonly contractorName?: string | null;
  readonly joiningDate?: string | null;
  readonly leavingDate?: string | null;
  readonly assignmentId: string;
  readonly farmId: string;
  readonly assignedFrom: string;
  readonly assignedTo?: string | null;
}

export type AttendanceEligibleWorkerList = PagedResponse<AttendanceEligibleWorker>;

export interface AttendanceRecord {
  readonly id: string;
  readonly organizationId: string;
  readonly farmId: string;
  readonly farmName: string;
  readonly workerId: string;
  readonly workerDisplayName: string;
  readonly workerFirstName: string;
  readonly workerLastName?: string | null;
  readonly gender: string;
  readonly laborCategoryId?: string | null;
  readonly laborCategoryName?: string | null;
  readonly employmentType: string;
  readonly attendanceDate: string;
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
  readonly calculatedRate?: number | null;
  readonly calculatedAmount?: number | null;
  readonly currencyId?: string | null;
  readonly currencyCode?: string | null;
  readonly currencySymbol?: string | null;
  readonly status: AttendanceStatus | string;
  readonly notes?: string | null;
  readonly finalizedAt?: string | null;
  readonly finalizedBy?: string | null;
  readonly createdAt: string;
  readonly createdBy: string;
  readonly updatedAt?: string | null;
  readonly updatedBy?: string | null;
}

export interface DailyAttendanceSummary {
  readonly totalCount: number;
  readonly workedCount: number;
  readonly fullDayCount: number;
  readonly halfDayCount: number;
  readonly hourlyCount: number;
  readonly notWorkedCount: number;
  readonly estimatedEarnings: number;
  readonly status: AttendanceStatus | string;
}

export interface DailyAttendanceResponse {
  readonly farmId: string;
  readonly farmName: string;
  readonly attendanceDate: string;
  readonly summary: DailyAttendanceSummary;
  readonly records: readonly AttendanceRecord[];
}

export interface DailyDraftAttendanceItemRequest {
  readonly id?: string | null;
  readonly workerId: string;
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
  readonly notes?: string | null;
}

export interface SaveDailyDraftAttendanceBatchRequest {
  readonly farmId: string;
  readonly attendanceDate: string;
  readonly items: readonly DailyDraftAttendanceItemRequest[];
  readonly removeOmittedDrafts?: boolean;
}

export interface CreateDraftAttendanceRequest {
  readonly farmId: string;
  readonly workerId: string;
  readonly attendanceDate: string;
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
  readonly notes?: string | null;
}

export interface UpdateDraftAttendanceRequest {
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
  readonly notes?: string | null;
}

export interface AttendanceWagePreviewRequest {
  readonly workerId: string;
  readonly attendanceDate: string;
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
  readonly farmId?: string | null;
}

export interface AttendanceWagePreviewResponse {
  readonly workerId: string;
  readonly workerDisplayName: string;
  readonly gender: string;
  readonly attendanceDate: string;
  readonly attendanceType: string;
  readonly resolvedWageType?: string | null;
  readonly wageType?: string | null;
  readonly rate: number;
  readonly quantity: number;
  readonly workingHours?: number | null;
  readonly calculatedAmount: number;
  readonly currencyId?: string | null;
  readonly currencyCode: string;
  readonly currencySymbol: string;
  readonly currency: string;
  readonly isEarningEligible: boolean;
  readonly isWorkerEligible: boolean;
  readonly ineligibilityReason?: string | null;
}

export interface AttendanceWagePreviewBatchItemRequest {
  readonly workerId: string;
  readonly attendanceType: AttendanceType | string;
  readonly workingHours?: number | null;
}

export interface AttendanceWagePreviewBatchRequest {
  readonly attendanceDate: string;
  readonly items: readonly AttendanceWagePreviewBatchItemRequest[];
  readonly farmId?: string | null;
}

export interface AttendanceWagePreviewBatchResponse {
  readonly attendanceDate: string;
  readonly totalCount: number;
  readonly workedCount: number;
  readonly fullDayCount: number;
  readonly halfDayCount: number;
  readonly hourlyCount: number;
  readonly notWorkedCount: number;
  readonly totalEstimatedEarnings: number;
  readonly items: readonly AttendanceWagePreviewResponse[];
}

export interface FinalizeAttendanceRequest {
  readonly farmId: string;
  readonly attendanceDate: string;
  readonly attendanceIds?: readonly string[] | null;
}

export interface FinalizeAttendanceResponse {
  readonly farmId: string;
  readonly farmName: string;
  readonly attendanceDate: string;
  readonly finalizedCount: number;
  readonly paidCount: number;
  readonly notWorkedCount: number;
  readonly totalEarnings: number;
  readonly summary: DailyAttendanceSummary;
  readonly records: readonly AttendanceRecord[];
}

export interface AttendanceGridRow {
  readonly id?: string | null;
  readonly workerId: string;
  readonly workerDisplayName: string;
  readonly workerFirstName: string;
  readonly workerLastName?: string | null;
  readonly gender: string;
  readonly mobileNumber?: string | null;
  readonly laborCategoryName?: string | null;
  readonly contractorName?: string | null;
  readonly employmentType: string;
  attendanceType: AttendanceType;
  workingHours?: number | null;
  calculatedRate?: number | null;
  calculatedAmount?: number | null;
  currencySymbol?: string | null;
  status: AttendanceStatus;
  notes?: string | null;
  isModified?: boolean;
}

export const ATTENDANCE_TYPE_OPTIONS: readonly {
  readonly value: AttendanceType;
  readonly label: string;
  readonly description: string;
}[] = [
  { value: "FULL_DAY", label: "Full Day", description: "Standard full day rate" },
  { value: "HALF_DAY", label: "Half Day", description: "Standard half day rate" },
  { value: "HOURLY", label: "Hourly", description: "Rate multiplied by hours" },
  { value: "NOT_WORKED", label: "Not Worked", description: "No earnings generated" },
];

export function formatAttendanceType(type?: string | null): string {
  switch (type?.toUpperCase()) {
    case "FULL_DAY":
      return "Full Day";
    case "HALF_DAY":
      return "Half Day";
    case "HOURLY":
      return "Hourly";
    case "NOT_WORKED":
      return "Not Worked";
    default:
      return type || "—";
  }
}

export function formatAttendanceStatus(status?: string | null): string {
  switch (status?.toUpperCase()) {
    case "DRAFT":
      return "Draft";
    case "FINALIZED":
      return "Finalized";
    default:
      return status || "—";
  }
}

export function getAttendanceTypeBadgeClass(type?: string | null): string {
  switch (type?.toUpperCase()) {
    case "FULL_DAY":
      return "type-badge-full-day";
    case "HALF_DAY":
      return "type-badge-half-day";
    case "HOURLY":
      return "type-badge-hourly";
    case "NOT_WORKED":
      return "type-badge-not-worked";
    default:
      return "type-badge-default";
  }
}

File: backend/src/FarmManagement.Application/DTOs/Labor/AttendanceModels.cs
namespace FarmManagement.Application.DTOs.Labor;

public sealed record AttendanceActor(Guid UserId, Guid OrganizationId);

public sealed record AttendanceEligibleWorkerResponse(
    Guid WorkerId,
    string DisplayName,
    string FirstName,
    string? LastName,
    string Gender,
    string EmploymentType,
    string? MobileNumber,
    Guid? LaborCategoryId,
    string? LaborCategoryName,
    Guid? ContractorId,
    string? ContractorName,
    DateOnly? JoiningDate,
    DateOnly? LeavingDate,
    Guid AssignmentId,
    Guid FarmId,
    DateOnly AssignedFrom,
    DateOnly? AssignedTo);

public sealed record CreateDraftAttendanceRequest(
    Guid FarmId,
    Guid WorkerId,
    DateOnly AttendanceDate,
    string AttendanceType,
    decimal? WorkingHours = null,
    string? Notes = null);

public sealed record UpdateDraftAttendanceRequest(
    string AttendanceType,
    decimal? WorkingHours = null,
    string? Notes = null);

public sealed record DailyDraftAttendanceItemRequest(
    Guid? Id,
    Guid WorkerId,
    string AttendanceType,
    decimal? WorkingHours = null,
    string? Notes = null);

public sealed record SaveDailyDraftAttendanceBatchRequest(
    Guid FarmId,
    DateOnly AttendanceDate,
    IReadOnlyList<DailyDraftAttendanceItemRequest> Items,
    bool RemoveOmittedDrafts = false);

public sealed record AttendanceRecordResponse(
    Guid Id,
    Guid OrganizationId,
    Guid FarmId,
    string FarmName,
    Guid WorkerId,
    string WorkerDisplayName,
    string WorkerFirstName,
    string? WorkerLastName,
    string Gender,
    Guid? LaborCategoryId,
    string? LaborCategoryName,
    string EmploymentType,
    DateOnly AttendanceDate,
    string AttendanceType,
    decimal? WorkingHours,
    decimal? CalculatedRate,
    decimal? CalculatedAmount,
    Guid? CurrencyId,
    string? CurrencyCode,
    string? CurrencySymbol,
    string Status,
    string? Notes,
    DateTimeOffset? FinalizedAt,
    Guid? FinalizedBy,
    DateTimeOffset CreatedAt,
    Guid CreatedBy,
    DateTimeOffset? UpdatedAt,
    Guid? UpdatedBy);

public sealed record DailyAttendanceSummaryResponse(
    int TotalCount,
    int WorkedCount,
    int FullDayCount,
    int HalfDayCount,
    int HourlyCount,
    int NotWorkedCount,
    decimal EstimatedEarnings,
    string Status);

public sealed record DailyAttendanceResponse(
    Guid FarmId,
    string FarmName,
    DateOnly AttendanceDate,
    DailyAttendanceSummaryResponse Summary,
    IReadOnlyList<AttendanceRecordResponse> Records);

public sealed record AttendanceWagePreviewRequest(
    Guid WorkerId,
    DateOnly AttendanceDate,
    string AttendanceType,
    decimal? WorkingHours = null,
    Guid? FarmId = null);

public sealed record AttendanceWagePreviewResponse(
    Guid WorkerId,
    string WorkerDisplayName,
    string Gender,
    DateOnly AttendanceDate,
    string AttendanceType,
    string? ResolvedWageType,
    string? WageType,
    decimal Rate,
    decimal Quantity,
    decimal? WorkingHours,
    decimal CalculatedAmount,
    Guid? CurrencyId,
    string CurrencyCode,
    string CurrencySymbol,
    string Currency,
    bool IsEarningEligible,
    bool IsWorkerEligible,
    string? IneligibilityReason = null);

public sealed record AttendanceWagePreviewBatchItemRequest(
    Guid WorkerId,
    string AttendanceType,
    decimal? WorkingHours = null);

public sealed record AttendanceWagePreviewBatchRequest(
    DateOnly AttendanceDate,
    IReadOnlyList<AttendanceWagePreviewBatchItemRequest> Items,
    Guid? FarmId = null);

public sealed record AttendanceWagePreviewBatchResponse(
    DateOnly AttendanceDate,
    int TotalCount,
    int WorkedCount,
    int FullDayCount,
    int HalfDayCount,
    int HourlyCount,
    int NotWorkedCount,
    decimal TotalEstimatedEarnings,
    IReadOnlyList<AttendanceWagePreviewResponse> Items);

public sealed record FinalizeAttendanceRequest(
    Guid FarmId,
    DateOnly AttendanceDate,
    IReadOnlyList<Guid>? AttendanceIds = null);

public sealed record FinalizeAttendanceResponse(
    Guid FarmId,
    string FarmName,
    DateOnly AttendanceDate,
    int FinalizedCount,
    int PaidCount,
    int NotWorkedCount,
    decimal TotalEarnings,
    DailyAttendanceSummaryResponse Summary,
    IReadOnlyList<AttendanceRecordResponse> Records);

public sealed record ExcludedWorkerInfo(
    Guid WorkerId,
    string DisplayName,
    string Reason);

public sealed record CopyPreviousDayPreviewWorkerItem(
    Guid WorkerId,
    string DisplayName,
    string FirstName,
    string? LastName,
    string Gender,
    string? MobileNumber,
    string? LaborCategoryName,
    string? ContractorName,
    string EmploymentType,
    string AttendanceType,
    decimal? WorkingHours,
    string? Notes,
    bool IsEligible,
    string? IneligibilityReason,
    decimal? ProvisionalRate,
    decimal? ProvisionalAmount);

public sealed record CopyPreviousDayPreviewResponse(
    Guid FarmId,
    string FarmName,
    DateOnly SourceDate,
    DateOnly TargetDate,
    int TotalSourceCount,
    int EligibleCount,
    int ExcludedCount,
    bool TargetHasExistingRecords,
    int TargetExistingRecordCount,
    bool TargetIsFinalized,
    DailyAttendanceSummaryResponse? SourceSummary,
    IReadOnlyList<CopyPreviousDayPreviewWorkerItem> Workers,
    IReadOnlyList<ExcludedWorkerInfo> ExcludedWorkers);

public sealed record CopyPreviousDayAttendanceRequest(
    Guid FarmId,
    DateOnly TargetDate,
    DateOnly? SourceDate = null,
    IReadOnlyList<Guid>? WorkerIds = null,
    bool OverwriteExistingDraft = false);

public sealed record CopyPreviousDayAttendanceResponse(
    Guid FarmId,
    string FarmName,
    DateOnly SourceDate,
    DateOnly TargetDate,
    int TotalSourceWorkers,
    int CopiedCount,
    int ExcludedCount,
    IReadOnlyList<ExcludedWorkerInfo> ExcludedWorkers,
    DailyAttendanceResponse DailyAttendance);

public sealed record RelatedEarningsLedgerResponse(
    Guid Id,
    Guid WorkerId,
    DateOnly EarningsDate,
    string WageType,
    decimal Quantity,
    decimal WageRate,
    decimal GrossAmount,
    Guid CurrencyId,
    string CurrencyCode,
    string CurrencySymbol,
    string Status,
    string EntryType,
    DateTimeOffset? FinalizedAt,
    Guid? FinalizedBy,
    string? Description);

public sealed record AttendanceDetailResponse(
    Guid Id,
    Guid OrganizationId,
    Guid FarmId,
    string FarmName,
    Guid WorkerId,
    string WorkerDisplayName,
    string WorkerFirstName,
    string? WorkerLastName,
    string Gender,
    Guid? LaborCategoryId,
    string? LaborCategoryName,
    string EmploymentType,
    string? MobileNumber,
    DateOnly AttendanceDate,
    string AttendanceType,
    decimal? WorkingHours,
    decimal? CalculatedRate,
    decimal? CalculatedAmount,
    Guid? CurrencyId,
    string? CurrencyCode,
    string? CurrencySymbol,
    string Status,
    string? Notes,
    DateTimeOffset? FinalizedAt,
    Guid? FinalizedBy,
    DateTimeOffset CreatedAt,
    Guid CreatedBy,
    DateTimeOffset? UpdatedAt,
    Guid? UpdatedBy,
    RelatedEarningsLedgerResponse? EarningsLedger);

